### DISCUSSION TITLE
Pint supports VictorMetrics

### DISCUSSION THREAD

**Discussion by rakeshpb90** [cloudflare/pint#869] [General]
I would like to know whether Pint supports Victoriametrics. I'm receiving a 400 response code for the following API's  /api/v1/status/config and /api/v1/status/flags

**Comment by wonderg** [comment]
It seems the challenge with Pint and VictoriaMetrics originates from Pint's use of the Prometheus-specific endpoint `/api/v1/status/config` in `internal/promapi/config.go`, which VictoriaMetrics doesn't support. While I'm not a Go developer and cannot guarantee a solution, a potential workaround might involve mapping the required configuration details (like `ExternalLabels`, `ScrapeInterval`, `ScrapeTimeout`, and `EvaluationInterval`) from equivalent VictoriaMetrics and VMAlert settings. This suggestion is speculative and would need further exploration to implement effectively.

**Comment by andrew-shevchyk** [comment]
I'm working on research how to implement VM support in pint (but I'm not Go developer either so using this opportunity to get familiar with it), I mocked responses from Prom server but faced another issue - VictoriaMetrics extends PromQL with additional functions so if an alerting rule uses them then it can't be parsed and the validation fails.
I'm looking into using [VM library](https://pkg.go.dev/github.com/VictoriaMetrics/metricsql) for parsing but not sure if it would be feasible to adapt due to significant differences in structures comparing to [PromQL parser](https://pkg.go.dev/github.com/prometheus/prometheus/promql/parser)

**Reply by wonderg** [reply]
Yes, you are right about MetricsQL, but it is probably worth starting with PromQL support only, because the reality is that only a few products use MetricsQL, while VictoriaMetrics is quite popular as a long term storage for metrics. So it is good to have support for MetricsQL, but not necessary, and in most cases PromQL support should be enough.

**Reply by antoinedeschenes** [reply]
Regarding compatibility, I've successfully disabled pint's `promql/syntax` check and combined `vmalert-prod -dryRun -rule 'rulefile.yaml'` to handle expr linting. 

I've also used yq to remove VM-only fields in the rules specification (ex. group-level `labels`, `concurrency`, etc.)

Now where I'm hitting problems is that vmalert supports [additional template values and functions](https://docs.victoriametrics.com/vmalert/#templating) over Prometheus in alert labels and annotations, and linting these expressions isn't covered by vmalert's validation checks.

**Reply by salarali** [reply]
@andrew-shevchyk were you able to enable victoria support?

**Reply by rcowham** [reply]
I have implemented the basics for VictoraMetrics support: https://github.com/rcowham/pint/blob/vm_test/docs/examples/victoriametrics.hcl#L8

It falls back to the VM parser (MetricsQL) if the Prometheus parser fails. Also validates extra VM template functions.

So if you configure it, you get warnings like:

```
pint --config ./pint_vm.hcl --show-duplicates lint perforce_rules.yml
level=INFO msg="Loading configuration file" path=./pint_vm.hcl
level=INFO msg="Finding all rules to check" paths=["perforce_rules.yml"]
level=INFO msg="Configured new Prometheus server" name=monitor uris=1 uptime=up tags=[] include=[] exclude=[]
level=INFO msg="Checking Prometheus rules" entries=45 workers=10 online=true
level=INFO msg="VictoriaMetrics mode enabled, PromQL syntax errors will be reported as warnings"
Warning: MetricsQL-specific syntax detected (promql/syntax)
  ---> perforce_rules.yml:100-101 -> `Vmagent not responding`
100 |         lag(node_time_seconds[24h]) > 5 * 60
              ^^^ unknown function with name "lag"
101 |             and on(instance) (p4_ra_instance_active > 0)

Warning: MetricsQL-specific syntax detected (promql/syntax)
  ---> perforce_rules.yml:114-115 -> `Vmagent not responding (Critical Care)`
114 |         lag(node_time_seconds[24h]) > 5 * 60
              ^^^ unknown function with name "lag"
115 |             and on(instance) (p4_ra_instance_active > 0 and p4_ra_critical_care == 1)

level=INFO msg="Problems found" Warning=2
```

Such warnings can be ignored. Although am considering that if you turn this on you may not even want the warning if it is valid VM syntax - thoughts?

Happy to look at a PR if of interest?

**Reply by rcowham** [reply]
@andrew-shevchyk @salarali FYI ^^

Will look at adding extra VM templating variables too (although it supports things like " has been true for {{ .For }}" already...

**Reply by rcowham** [reply]
Could have a look at avoiding /status/config etc too - although these can be avoided with a config file entry:

```
# Disable checks which don't work with Victoria Metrics (lacks config and metadata endpoints)
checks {
  disabled = ["promql/rate", "promql/range_query"]
}
```

**Comment by wbollock** [comment]
Looks like this recent release could help as unsupported endpoints get ignored now https://github.com/cloudflare/pint/releases/tag/v0.70.0
